var/spack/repos/builtin/packages/glib/package.py

```diff
@@ -53,6 +53,8 @@ class Glib(AutotoolsPackage):
     # Clang doesn't seem to acknowledge the pragma lines to disable the -Werror
     # around a legitimate usage.
     patch('no-Werror=format-security.patch')
+    # GCC8 on mac can preprocess obj-c headers but not build them
+    patch('configure-darwin.patch', when='platform=darwin')
```

Comment:
Should this only apply for `gcc` or `gcc@8:`, based on the above comment?

Also to check: did you create this patch yourself? If not, it would be worth adding a link since that helps keep track of when it is no longer needed. (this applies to other patches added in the PR)

Reply:
I created the patch myself. I believe it's more correct than what existed in the configure file -- the compiler *will* have to build the frameworks eventually, so checking to see that the carbon/cocoa headers exist may not be sufficient. I don't know if other GCC versions (or other non-apple compilers) can build on the mac without the patch, but I suspect not.

Reply:
>  the compiler will have to build the frameworks eventually, so checking to see that the carbon/cocoa headers exist may not be sufficient

True although I wonder if in general that kind of check would be hindered because the resources to truly build are not available at the time (so the cpp serves as a sanity check to resolve the includes).

Is this a similar issue to https://github.com/spack/spack/pull/10944#discussion_r270608085 where glib assumes the presence of some libraries based on include files but then cannot compile? If so, is there a reason you made the check more stringent here vs. the approach for python of adding the appropriate dependency?

Reply:
No, the other case was a link failure. This failure is that the downstream code has valid syntax (preprocesses fine) but has *semantics* that are incompatible with GCC8:
```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/System/Library/Frameworks/Security.framework/Headers/Authorization.h:193:7: error: variably modified 'bytes' at file scope
  char bytes[kAuthorizationExternalFormLength];
```

Reply:
If I understand correctly, the affect is to disable these features (e.g. `glib_have_cocoa`) when building with gcc, so I would like to make sure: do you know if these are enabled when you build the package with apple's clang?

Reply:
I just reconfigured with and without the patch with clang and gcc. Without the patch, clang successfully detected and built both cocoa and carbon. With the patch, clang only built carbon.
Without the patch, gcc configure says it can build carbon and cocoa, but the compiler errors when it includes the cocoa framework. With the patch, gcc configure says it can build neither carbon nor cocoa.

It sounds like I should change the patch command to
```py
    patch('configure-darwin.patch', when='%gcc platform=darwin')
```
to preserve the original behavior for the Apple clang compiler?

Reply:
> It sounds like I should change the patch command to...

@sethrj I agree. Based on your results, I'm guessing that `if ac_fn_c_try_compile` is too strict of a test for `cocoa` (whether clang or gcc is used) so you may want to relax that.

This brings up another point though: when it comes to reviewing Spack packages, I'd like to avoid reviewing novel patches for code when possible. Sometimes that's unavoidable but when that happens, if you could attach as much info as possible in the comment so other users of the package can track it that would be great:

* A link to any open issues regarding the failure (either in or outside of Spack)
* A description of the details of the system/compiler in used (e.g. including compiler version, even if the patch `when` clause is broader)
* The error message you get that this resolves